Demo/FTP/client.cpp

```diff
@@ -0,0 +1,254 @@
+            while (std::getline(pss, token, ',')) {
+                try {
+                    nums.push_back(std::stoi(token));
+                } catch (...) {
+                    std::cerr << "Invalid PORT parameters.\n";
+                    nums.clear();
+                    break;
+                }
+            }
+            if (nums.size() != 6) {
+                std::cerr << "PORT command requires 6 comma-separated values.\n";
+                continue;
+            }
+
+            std::string ipStr = std::to_string(nums[0]) + "." + std::to_string(nums[1]) + "." +
+                                std::to_string(nums[2]) + "." + std::to_string(nums[3]);
+            int dataPort = nums[4] * 256 + nums[5];
+
+            if (dataListenerSock != -1) {
+                close(dataListenerSock);
+                dataListenerSock = -1;
+            }
+
+            dataListenerSock = socket(AF_INET, SOCK_STREAM, 0);
+            if (dataListenerSock < 0) {
+                std::cerr << "Failed to create data listener socket: " << strerror(errno) << "\n";
+                continue;
+            }
+
+            sockaddr_in dataAddr{};
+            dataAddr.sin_family = AF_INET;
+            dataAddr.sin_port = htons(dataPort);
+            dataAddr.sin_addr.s_addr = INADDR_ANY;
+
+            int opt = 1;
+            setsockopt(dataListenerSock, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt));
+
+            if (bind(dataListenerSock, (sockaddr*)&dataAddr, sizeof(dataAddr)) < 0) {
+                std::cerr << "Failed to bind data listener socket: " << strerror(errno) << "\n";
+                close(dataListenerSock);
+                dataListenerSock = -1;
+                continue;
+            }
+
+            if (listen(dataListenerSock, 1) < 0) {
+                std::cerr << "Failed to listen on data listener socket: " << strerror(errno) << "\n";
+                close(dataListenerSock);
+                dataListenerSock = -1;
+                continue;
+            }
+
+            portSet = true;
+
+            std::string portCmd = "PORT " + params + "\r\n";
+            if (!sendAll(ctrlSock, portCmd)) {
+                std::cerr << "Failed to send PORT command\n";
+                break;
+            }
+
+            if (readLine(ctrlSock, line) > 0)
+                std::cout << line << "\n";
+
+        } else if (cmd == "LIST" || cmd == "RETR" || cmd == "STOR") {
+            if (!portSet || dataListenerSock == -1) {
+                std::cerr << "You must issue a PORT command before " << cmd << "\n";
+                continue;
+            }
+
+            if (!sendAll(ctrlSock, inputLine + "\r\n")) {
+                std::cerr << "Failed to send " << cmd << " command\n";
+                break;
+            }
+
+            sockaddr_in serverDataAddr{};
+            socklen_t serverDataLen = sizeof(serverDataAddr);
+            int dataSock = accept(dataListenerSock, (sockaddr*)&serverDataAddr, &serverDataLen);
```

Comment:
This line means the client is acting like a server - it’s passively accepting a connection.
Wrap `accept()` in `poll()` with timeout to avoid hanging indefinitely if the server never connects.